man/gcin2tab.1
.TH GCIN 1 "21 JAN 2008" "GCIN 1.3.8" "gcin input method platform"
.SH NAME
gcin2tab \- converts text input method table to gcin's .gtab format
.SH SYNOPSIS
.B gcin2tab
.B [\fB-v\fP|\fB--version\fP]
.br
.SH DESCRIPTION
.B gcin2tab
converts plain text input method table to gcin's .gtab binary format.
.SH OPTIONS
.TP
.B \-v, \-\-version
Show version of gcin2tab.
.SH SEE ALSO
.BR gcin (1).
.SH AUTHOR
gcin2tab was written by Edward Der-Hua Liu <cp76@csie.nctu.edu.tw>.


Changelog
** UTF-8 encoding **
1.4.3:
- 詞音 Firefox 用 Shift 切到英數仍有 window 跑出來的問題。
- 修正許氏鍵盤無法輸入 ,./ 的問題。
- 修正 gdm 輸入密碼錯誤時 greeter coredump 的問題。
- gtab 輸入法由詞庫自動選字改成與詞音一樣用輸入緩衝區的方式。輸入法如嘸蝦米等按空白自動送出第一字的就不支援，如果想要用，可以在 gcin-setup 設定空白鍵的選項。
- qt4 IM module 的支援，skype 上也可以用 Shift 切換中英了
- gcin-setup help 改成用 coolcd 編輯過的 html, 感謝 coolcd
- gcin 現在支援 su root 的中文輸入，必須配合最新的 libgcin-im-client.so.1.1.0
- 詞音在編輯區是空的時候按 Shift+xxxx 鍵輸入標點符號，回直接送出不會放到編輯區。

1.4.2:
- 詞音空白鍵選項設定成輸入空白時，選字會選錯的問題。
- 詞音使用 CapsLock 切換中英時，大小寫不對的問題。

1.4.1:
- 現在可以用 mouse 在預選詞的視窗點選，主要給小朋友用。
- 中文輸入狀態按 Shfit + abc  改成輸入大寫。
- tray 上面的選單現在可以切換是否要念出發音 & 也可以切換小鍵盤。
- firefox 3.0/gnome-terminal 關閉 coredump 的問題。

1.4.0:
- 按 keypad . 現在可以輸入 '. '了 。感謝jackblackevo的建議。
- ctrl-alt-, 有時候需要按兩次才會出現的問題。
- tray panel 如果放在最上面，小鍵盤位置會蓋掉 tray 的問題。
- 詞音在某些情況 coredump 的 bug。
- 符號視窗按滾輪捲動後會自動關閉的問題。
- Cheng-Wei Lin提供的：gcin-setup 選項可以設定是否要使用 詞音 ctrl-E 編輯模式。內定是關閉。
- gtab 大小寫如果有不同的定義，現在螢幕小鍵盤會顯示不同的 keyname
- 符號輸入法，一般使用的方法為加入 Ctrl-Shift 循環內，進入符號輸入法後螢幕小鍵盤會自動顯示，按任一鍵輸入後，會自動切回內定的輸入法，同時關閉螢幕小鍵盤。直接按 Ctrl-Alt-4(原標點倉頡) 進入符號輸入也可以。感謝 winlin 的協助。
- 使用 CapsLock 輸入英數時，不能輸入 `*? 的問題。感謝 coolcd 的 bug report.
- 詞音內部的英數模式現在在彈出式的視窗模式現在不會跑出 window 了。
- tsa2d32 Bus alignment 的問題(如 sparc)。
- 符號視窗改用 coolcd 提供的 symbol-table
- Alt-Shift ←→↓↑ key 現在也可以定義 key 了，default 就是定義成 ←→↓↑。
- gtab 輸入法按滿送出字後，按空白鍵的問題。
- 倚天26鍵注音鍵盤數字鍵沒辦法輸入的問題。
- `? 反引號輸入符號的方式改成與新注音類似。感謝 coolcd 提供的 table
- 詞音斷詞選字的 bug

1.3.9:
- 改用自己的简←→正體字轉換，這樣 sim2trad 就不會有掉字的情形。感謝 LGJ 的協助。
- gtab 輸入法 coredump的 bug

1.3.8:
- 詞音按 Delete/Backspace 把編輯區的內容清光後，現在會自動把 Window 隱藏起來。
-注音【】現在可用來輸入 全形（）。
- gtab 現在支援快碼輸入法
- 詞音在用 Shift 預選詞後不會斷詞選詞的問題。
- Ctrl-Shift-;   現在可以用來輸入全形'：'，不過只限非 XIM 的情況。
- 詞音在中文狀態 Shift Enter 可以用了。
- 在使用彈出是輸入視窗時，符號輸入視窗顯示的問題。
- 新增點選符號視窗裡的按鈕後自動關閉的選項，內定是開啟，可以用 gcin-setup 關閉。
- winlin 提供的新 icons
- 現在可以設定 gtab/詞音 選字 window 裏面選擇鍵的顏色。
- 修正 gtab 的 bug，現在應該可以定義 63 個 keys (沒有實際測過)
- sim2trad 如果有特殊符號就無法翻譯的問題。
- Cheng-Wei Lin提供的：新增了一個緩衝區編輯模式，使用者可以在詞音輸入法下按下 ctrl+e 切換緩衝區編輯模式；在緩衝區編輯模式下，可以用 h 跟 l 來左右移動，按下 x 則可刪除。 
- 提供小鍵盤，可控制所有X window 的程式，不限於輸入法的部份。在gcin tray icon(圖像) 按滑鼠中鍵。此功能放在 libgcin1.so，有用到才會 load 到 memory。也可以用 gcin-kbm-toggle 這個可執行檔切換。
- Big endian CPU 上可能出現的 bug.
- 使用簡體字輸出時，如果有無法轉換的情形，現在改成輸出原來的而不是忽略。

1.3.7:
- 有選項可以設定直接進入中文輸入的狀態，不過只能用在非 XIM 的情況。
- 1.3.6 tray 後面會空白的 bug
- Add Netscape Navigator 9 to gtk-im. Thanks to swyear

1.3.6:
- 詞音斷詞選詞的 bug，造成無法手動斷詞及嚴重影響正確率。
- 當詞音空白鍵選項為「輸入空白」，則右邊的 keypad 數字區可以用來選擇同音字；如果是選擇同音字，則是用來輸入數字。
- 詞音同音字選擇，中文字不會對齊的問題。
- 修正smartboy (小光光)回報 tsin.src 裏面資料的錯誤。
- tray icon 不見的問題，特別是 icewm，目前的解法會造成 gcin tray icon 晚5 秒鐘出現。

1.3.5:
- gcin 現在輸入注音 gtab輸入法時可以念出正在輸入的字，必須安裝語音資料庫於 /usr/share/gcin/ogg or /usr/local/share/gcin/ogg，還需要 ogg123 這個執行檔。
- gcin-setup po fix from caleb
- 新增在英數狀態是否要使用 alt-shift 片語輸入的選項。
- 詞音如果遇到同音詞，且這些同音詞都還沒用過，不會選第一個，造成不方便。
- 提供 Windows ctrl-,./;'  輸入標點符號的方式，不過 '：' 由於會與 ctrl-shift 衝，所以沒辦法用，目前暫時不想去理，反正替代的方式太多了。定義檔為  phrase-ctrl.table，必須中文輸入狀態才可以使用。
- more i18n fix from caleb
- 由某人提供的  --use_tsin=N 選項，可以把詞音的支援拿掉。
- 由 caleb 提供的 gcin-setup-tab
- 符號視窗不會顯示 gtab 輸入法按鍵 tip 的問題。
- 新的 mozilla 家族
- gtab 輸入法如果有比目前更短的拆法，拆字顯示區的顏色會變成黃色。
- gtab 新增是否要用?萬用字元的選項。
- 詞音空白鍵選項中的「送出編輯區內容」選項改成「輸入空白」。

1.3.4:
- 由於如果使用彈出式(popup) 輸入視窗，其實只有在顯示按鍵的時候需要移動視窗，這樣可以減少gcin在 firefox 一類 mozilla 家族程式的奇怪問題。現在也把popup 改成內定的選擇。用來用去，還是比較習慣不用 popup 的方式，不用一直看 tray 是中/英，比較不會按錯。
- gtab 如果選用不讓錯誤鍵進入(非傳統)，按下 space 時是無效的拆字，現在可以直接輸入下一個字，不用按 Escape 清除。
- 「唯一選擇時自動送出」這個選項在某些時候會造成 coredump 的問題。
- \ 現在可以定義輸入法了。
- gtab 輸入法選擇不允許錯誤鍵進入，按空白鍵出現紅色警告字不存在後，可以按空白鍵清除，也可以繼續按下一個未輸入的鍵，紅色也會清除。
-  詞音在按 CapsLock 輸入英文後有時候會造成前面已經對的詞後來又變錯的。
- FREEBSD 用 GTK_IM_MODULE=gcin 無法輸入英文的問題。
- gcin-setup 新增結束 gcin 的功能，只有從 gcin tray menu 內叫出 gcin-setup 才可以看到這個選項。
- gcin 如果重新執行，在之前執行的firefox 內輸入文字，gcin 的視窗就不會自動拉到最上面的問題。需更新 im-client。
- man page update from caleb.
- 由 winlin 提供的 icon files。
- gtab 如果定義的字串裏面含有 '<'，則會出現底下的訊息且無法顯示。Gtk-WARNING **: Failed to set text from markup due to error
- 純注音(ctrl-alt-3)現在也支援拼音了，所以現在可以用拼音注音查詢查詢其他的輸入法。
- build fix from Funda Wang.
- 由 caleb 提供的 gcin-setup i18n po patch，由於會對某些人造成困擾而且大部份的人不需要，configure 需要 --use_i18n=Y啟用。
- gtab 不允定義 & 的問題。
- 新增 詞音/gtab CapsLock英數狀態大小寫的選項。

1.3.3:
- 詞音新增同音字詞選擇鍵在後的選項，這樣做的好處是當你由左到右看到某個字詞符合後，就不用再回頭看是那一個鍵。
- 詞音Tab 斷詞改成可以在任意地方按，代表詞從這邊開頭，如果在結尾按則代表把最後一個字當開頭。例如輸入「他示意蜘蛛」，我們想要的是「他是一隻豬」，這時直接在結尾按 Tab 一次就可以了，代表把豬當成一個單字的詞。當使用者自己按空白鍵選詞時，會把這個手動斷詞清除。不過很可惜的，在某些程式不能用，例如 gedit，因為 gedit沒有把  Tab key 傳給 gcin 處理。為了解決這個問題，也可以改按 Escape。
- 詞音編輯區如果沒有東西，這時按 escape 鍵 gcin 會傳回不處理，所以 AP 會自己處理 escape 鍵。
- 在選字詞 menu 出現時按 escape 應該只把 menu 關閉，不要斷詞。
- gcin2tab 現在遇到沒有定義 keyname 的鍵會中斷而不是忽略。
- 詞音「試試新的」-> [是世新的]，由於單字詞「的」頻率出現次數過高，會造成錯誤，把較長斷詞的比重調高。
- gtab 選擇頁數號碼不會清除的問題。
- gtab 在只有顯示輸入鍵的情況，輸入視窗會超出螢幕的問題。
- gtab 垂直選擇頁數顯示錯誤。
- 簡易輸入法的問題。
- 嘗試解決 gcin 在 beryl/firefox 上的問題。不確定是否有用，在 mandriva cooker 上似乎可以。
- 新增注音鍵盤「神通」。
- 由某位提供的 ./configure --use_tray=N，不想要 tray 時可以用此選項。
- alsa2 與 im-client 起衝突的問題，造成 flash player coredump。必須 update libgcin-im-client.so。
- 為了解決 flashplayer9 當掉的問題，XIM 的 IMdkit 換成 scim 的，現在不用 unset XMODIFIERS 似乎也不會當了。如果你必須經常重新啟動 gcin 最好還是在 firefox 的 start-script 裏面用 unset XMODIFIERS，因為新的 IMdkit 在 gcin 重新啟動有時候還是會造成正在執行的 firefox coredump。而且這樣可以節省 memory。
- gtab 唯一選擇時自動送出的新選項。例如我們在倉頡裏面定義「benq 明基」，當按到 ben 時，由於此時沒有其他的是符合 'ben*'，這時候就會自動送出。
- []\ 現在也可以用來切換輸入法了。已經把拉丁字母及帶調粵拼放上去。

1.3.2:
- 新增編輯 gtab 輸入法外加字詞的功能，可以在 gcin-setup -> gtab 中執行。也可以自己手動編輯 ~/.gcin/xxx.gtab.append，其中 xxx 是你內定的 .gtab 輸入法，例如使用倉頡的人其 filename 就是 ~/.gcin/cj.gtab.append。gcin 會在按 ctrl-space 時檢查 file 是否有被修改，自動 reload。
- candyz 提供的 gtab 垂直選字詞的方式，以解決如果詞很長，無法完全顯示的問題。
- gtab 新增可以用  page up/down or keypad +- 跳頁。對嘸蝦米等 space 送出第一個且定義超過10個時就可以用到。
- gcin2tab 現在不會刪除重複的定義，如果不喜歡，請刪除 .cin 重複定義的部份。
- 超過 5 個 key 的時候，相同拆字鍵的字不會保留原來順序的問題。
- 現在 freebsd 上應該是不用 mergesort 了。
- 因為發現有不少人沒有設定 locale 為 zh_TW，也沒有設定 GTK_IM_MODULE=gcin，造成 install 1.3.1 後無法使用 gcin。所以把 gtk im module 的 language 適用範圍改回 '*'，省掉這個麻煩。
- 如果使用注音輸入拼音，現在可以 juyin-learn 可以顯示成聲調拼音的注音方式。
- gtab 在顯示片語的時候，如果遇到 http:// ftp:// 會不顯示這一部份，現在也改成至少顯示 18個 bytes。
- 新增設定文字及視窗背景顏色的功能。不過取消還原成系統的 theme顏色 後不會馬上還原，必須 pkill gcin。
- noseeing-inst bugfix from LGJ
- 使用一般輸入及萬用字元(*?) 查詢現在會顯示目前及總共的頁數。

gcin2tab.c

/*
	Copyright (C) 1995-2008	Edward Der-Hua Liu, Hsin-Chu, Taiwan
*/

#include <stdio.h>
#include <stdarg.h>
#include <sys/types.h>
#if FREEBSD
#include <sys/param.h>
#include <sys/stat.h>
#endif
#include <string.h>
#include "gcin.h"
#include "gtab.h"
#include "gcin-endian.h"

FILE *fr, *fw;
int lineno;
char tt[512];


char *skip_spc(char *s)
{
	while ((*s==' ' || *s=='\t') && *s) s++;
	return s;
}

char *to_spc(char *s)
{
	while (*s!=' ' && *s!='\t' && *s) s++;
	return s;
}

void del_nl_spc(char *s)
{
  char *t;

  int len=strlen(s);
  if (!*s) return;

  t=s+len-1;

  while (*t=='\n' || *t==' ' || (*t=='\t' && t > s))
    t--;

  *(t+1)=0;
}


void get_line()
{
  while (!feof(fr)) {
    bzero(tt, sizeof(tt));
    fgets((char *)tt, 512, fr);
    lineno++;

    int len=strlen(tt);
    if (tt[len-1]=='\n')
      tt[len-1] = 0;

    if (tt[0]=='#' || strlen(tt) < 3)
      continue;
    else
      break;
  }
}

void cmd_arg(u_char **cmd, u_char **arg)
{
  char *t;

  get_line();
  char *s=tt;

  if (!*s) {
    *cmd=*arg=s;
    return;
  }

  s=skip_spc(s);
  t=to_spc(s);
  *cmd=s;
  if (!(*t)) {
    *arg=t;
    return;
  }

  *t=0;
  t++;

  t=skip_spc(t);
  del_nl_spc(t);

  char *p;
  if ((p=strchr(t, '\t')))
    *p = 0;

  *arg=t;
}

int sequ(char *s, char *t)
{
  return (!strcmp(s,t));
}

typedef struct {
  u_int key;
  u_char ch[CH_SZ];
  int oseq;
} ITEM2;

typedef struct {
  u_int64_t key;
  u_char ch[CH_SZ];
  int oseq;
} ITEM2_64;


#define MAX_K (500000)

ITEM2 itar[MAX_K];
ITEM2_64 itar64[MAX_K];

ITEM itout[MAX_K];
ITEM64 itout64[MAX_K];


int qcmp(const void *aa, const void *bb)
{
  ITEM2 *a = (ITEM2 *)aa, *b = (ITEM2 *) bb;

  if (a->key > b->key) return 1;
  if (a->key < b->key) return -1;

  return a->oseq - b->oseq;
}


int qcmp_64(const void *aa, const void *bb)
{
  ITEM2_64 *a = (ITEM2_64 *)aa, *b = (ITEM2_64 *) bb;

  if (a->key > b->key) return 1;
  if (a->key < b->key) return -1;

  return a->oseq - b->oseq;
}


#define mtolower(ch) (ch>='A'&&ch<='Z'?ch+0x20:ch)

static char kno[128];

int main(int argc, char **argv)
{
  int i;
  char fname[64];
  char fname_cin[64];
  char fname_tab[64];
  u_char *cmd, *arg;
  struct TableHead th;
  int KeyNum;
  char kname[128][CH_SZ];
  char keymap[128];
  int chno;
  gtab_idx1_t idx1[256];
  char def1[256];
  int quick_def;
  int *phridx=NULL, phr_cou=0;
  char *phrbuf = NULL;
  int prbf_cou=0;

  dbg("-- gcin2tab encoding UTF-8 --\n");
  dbg("--- please use iconv -f big5 -t utf-8 if your file is in big5 encoding\n");

  if (argc<=1) {
          printf("Enter table file name [.cin] : ");
          scanf("%s", fname);
  } else strcpy(fname,argv[1]);


  if (!strcmp(fname, "-v") || !strcmp(fname, "--version")) {
    dbg("gcin2tab for gcin " GCIN_VERSION "\n");
    exit(0);
  }

  char *p;
  if(p=strstr(fname, ".cin"))
    *p = 0;

  strcpy(fname_cin,fname);
  strcpy(fname_tab,fname);
  strcat(fname_cin,".cin");
  strcat(fname_tab,".gtab");

  if ((fr=fopen(fname_cin,"r"))==NULL)
          p_err("Cannot open %s\n", fname_cin);

  bzero(&th,sizeof(th));
  bzero(kno,sizeof(kno));
  bzero(keymap,sizeof(keymap));

  bzero(itar,sizeof(itar));
  bzero(itout,sizeof(itout));
  bzero(itar64,sizeof(itar64));
  bzero(itout64,sizeof(itout64));

  cmd_arg(&cmd, &arg);
  if (sequ(cmd, "%gen_inp")) {
    dbg("skip gen_inp\n");
    cmd_arg(&cmd, &arg);
  }

  if (!sequ(cmd,"%ename") || !(*arg) )
    p_err("%d:  %%ename english_name  expected", lineno);
  arg[15]=0;
//  strcpy(th.ename,arg);

  cmd_arg(&cmd, &arg);
  if (!(sequ(cmd,"%prompt") || sequ(cmd,"%cname")) || !(*arg) )
    p_err("%d:  %%prompt prompt_name  expected", lineno);
  strncpy(th.cname, arg, MAX_CNAME);
  printf("cname %s\n", th.cname);

  cmd_arg(&cmd, &arg);
  if (!sequ(cmd,"%selkey") || !(*arg) )
    p_err("%d:  %%selkey select_key_list expected", lineno);
  strcpy(th.selkey,arg);

  cmd_arg(&cmd, &arg);
  if (!sequ(cmd,"%dupsel") || !(*arg) ) {
    th.M_DUP_SEL = strlen(th.selkey);
  }
  else {
    th.M_DUP_SEL=atoi(arg);
    cmd_arg(&cmd, &arg);
  }

  if (sequ(cmd,"%endkey")) {
    strcpy(th.endkey, arg);
    cmd_arg(&cmd, &arg);
  }

  if (sequ(cmd,"%space_style")) {
    th.space_style = atoi(arg);
    cmd_arg(&cmd, &arg);
  }

  if (sequ(cmd,"%keep_key_case")) {
    th.flag |= FLAG_KEEP_KEY_CASE;
    cmd_arg(&cmd, &arg);
  }

  if (sequ(cmd,"%symbol_kbm")) {
    th.flag |= FLAG_GTAB_SYM_KBM;
    cmd_arg(&cmd, &arg);
  }

  if (sequ(cmd,"%phase_auto_skip_endkey")) {
    th.flag |= FLAG_PHRASE_AUTO_SKIP_ENDKEY;
    cmd_arg(&cmd, &arg);
  }

  if (!sequ(cmd,"%keyname") || !sequ(arg,"begin")) {
    p_err("%d:  %%keyname begin   expected, instead of %s %s", lineno, cmd, arg);
  }

  for(KeyNum=0;;) {
    char k;

    cmd_arg(&cmd, &arg);
    if (sequ(cmd,"%keyname")) break;
    if (BITON(th.flag, FLAG_KEEP_KEY_CASE))
      k=cmd[0];
    else
      k=mtolower(cmd[0]);

    if (kno[(int)k])
      p_err("%d:  key %c is already used",lineno, k);

    kno[(int)k]=++KeyNum;
    keymap[KeyNum]=k;
    bchcpy(&kname[KeyNum][0], arg);
  }

  keymap[0]=kname[0][0]=kname[0][1]=' ';
  KeyNum++;
  th.KeyS=KeyNum;    /* include space */

  cmd_arg(&cmd, &arg);

  if (sequ(cmd,"%quick") && sequ(arg,"begin")) {
    dbg(".. quick keys defined\n");
    for(quick_def=0;;) {
      char k;
      int len;

      cmd_arg(&cmd, &arg);
      if (sequ(cmd,"%quick")) break;
      k=kno[mtolower(cmd[0])]-1;

      int N = 0;
      char *p = arg;

      if (strlen(cmd)==1) {
        while (*p) {
          int len=u8cpy(th.qkeys.quick1[(int)k][N++], p);
          p+=len;
        }
      } else
      if (strlen(cmd)==2) {
        int k1=kno[mtolower(cmd[1])]-1;
        while (*p) {
          char tp[4];
          int len=u8cpy(tp, p);

          if (utf8_eq(tp,"□"))
             tp[0]=0;

          u8cpy(th.qkeys.quick2[(int)k][(int)k1][N++], tp);
          p+=len;
        }
      } else
        p_err("%d:  %quick only 1&2 keys are allowed '%s'", lineno, cmd);

      quick_def++;
    }
  }

  long pos=ftell(fr);
  int olineno = lineno;
  gboolean key64 = FALSE;

  while (!feof(fr)) {
    int len;

    cmd_arg(&cmd,&arg);
    if (!cmd[0] || !arg[0])
      continue;

    if (!strcmp(cmd, "%chardef")) {
      if (!strcmp(arg, "end"))
        break;
      else
        continue;
    }

    len=strlen(cmd);

    if (len > 5) {
      key64 = TRUE;
      dbg("more than 5 keys, will use 64 bit");
      break;
    }
  }

  fseek(fr, pos, SEEK_SET);
  lineno=olineno;

  INMD inmd, *cur_inmd = &inmd;

  cur_inmd->key64 = key64;
  cur_inmd->tbl64 = itout64;
  cur_inmd->tbl = itout;

  if (KeyNum < 64)
    cur_inmd->keybits = 6;
  else
    cur_inmd->keybits = 7;

  dbg("KeyNum:%d keybits:%d\n", KeyNum, cur_inmd->keybits);

  th.keybits = cur_inmd->keybits;
  cur_inmd->last_k_bitn = (((cur_inmd->key64 ? 64:32) / cur_inmd->keybits) - 1) * cur_inmd->keybits;


  puts("char def");
  chno=0;
  while (!feof(fr)) {
    int len;
    u_int64_t kk;
    int k;

    cmd_arg(&cmd, &arg);
    if (!cmd[0] || !arg[0])
      continue;

    if (!strcmp(cmd, "%chardef")) {
      if (!strcmp(arg, "end"))
        break;
      else
        continue;
    }

    len=strlen(cmd);

    if (len > th.MaxPress) {
      th.MaxPress=len;
    }

    if (len > 10)
      p_err("%d:  only <= 10 keys is allowed '%s'", lineno, cmd);

    kk=0;
    for(i=0;i<len;i++) {
      int key = BITON(th.flag, FLAG_KEEP_KEY_CASE) ?
        cmd[i] : mtolower(cmd[i]);

      k=kno[key];

      if (!k)
        p_err("%d: key undefined in keyname '%c'\n", lineno, cmd[i]);

      kk|=(u_int64_t)k << ( LAST_K_bitN - i*th.keybits);
    }

    if (key64) {
      memcpy(&itar64[chno].key, &kk, 8);
      itar64[chno].oseq=chno;
    }
    else {
      u_int key32 = kk;

      memcpy(&itar[chno].key, &key32, 4);
      itar[chno].oseq=chno;
    }

    if ((len=strlen(arg)) <= CH_SZ && (arg[0] & 0x80)) {
      char out[CH_SZ+1];

      bzero(out, sizeof(out));
      memcpy(out, arg, len);

      if (key64)
        bchcpy(itar64[chno].ch, out);
      else
        bchcpy(itar[chno].ch, out);

    } else {
      if (key64) {
          itar64[chno].ch[0]=phr_cou>>16;
          itar64[chno].ch[1]=(phr_cou >> 8) & 0xff;
          itar64[chno].ch[2]=phr_cou&0xff;
      }
      else {
          itar[chno].ch[0]=phr_cou>>16;
          itar[chno].ch[1]=(phr_cou >> 8) & 0xff;
          itar[chno].ch[2]=phr_cou&0xff;
      }

      if (len > MAX_CIN_PHR)
        p_err("phrase too long: %s  max:%d bytes\n", arg, MAX_CIN_PHR);

      phridx = trealloc(phridx, int, phr_cou+1);
      phridx[phr_cou++]=prbf_cou;
      phrbuf = realloc(phrbuf, prbf_cou + len + 1);
      strcpy(&phrbuf[prbf_cou],arg);
//      printf("phrase:%d  len:%d'%s'\n", phr_cou, len, arg);
      prbf_cou+=len;
    }

    chno++;
  }
  fclose(fr);

#define _sort qsort

  dbg("MaxPress: %d\n", th.MaxPress);

  th.DefC=chno;
  cur_inmd->DefChars = chno;

  if (key64)
    _sort(itar64,chno,sizeof(ITEM2_64),qcmp_64);
  else
    _sort(itar,chno,sizeof(ITEM2),qcmp);

  if (key64) {
    for(i=0;i<chno;i++)
      memcpy(&itout64[i],&itar64[i],sizeof(ITEM64));
  } else {
    for(i=0;i<chno;i++)
      memcpy(&itout[i],&itar[i],sizeof(ITEM));
  }


  bzero(def1,sizeof(def1));
  bzero(idx1,sizeof(idx1));


  u_int64_t keymask = KEY_MASK;
  for(i=0; i<chno; i++) {
    u_int64_t key = CONVT2(cur_inmd, i);
    int kk = (key>>LAST_K_bitN) & keymask;

    if (!def1[kk]) {
      idx1[kk]=(gtab_idx1_t)i;
      def1[kk]=1;
    }
  }

  idx1[KeyNum]=chno;
  for(i=KeyNum-1;i>0;i--)
    if (!def1[i]) idx1[i]=idx1[i+1];

  if ((fw=fopen(fname_tab,"w"))==NULL) {
    p_err("Cannot create");
  }

  printf("Defined Characters:%d\n", chno);

#if NEED_SWAP
  swap_byte_4(&th.version);
  swap_byte_4(&th.flag);
  swap_byte_4(&th.space_style);
  swap_byte_4(&th.KeyS);
  swap_byte_4(&th.MaxPress);
  swap_byte_4(&th.M_DUP_SEL);
  swap_byte_4(&th.DefC);
  for(i=0; i <= KeyNum; i++)
    swap_byte_4(&idx1[i]);
#endif
  fwrite(&th,1,sizeof(th),fw);
  fwrite(keymap, 1, KeyNum, fw);
  fwrite(kname, CH_SZ, KeyNum, fw);

  fwrite(idx1, sizeof(gtab_idx1_t), KeyNum+1, fw);


  if (key64) {
#if NEED_SWAP
    for(i=0; i < chno; i++) {
      swap_byte_8(&itout64[i].key);
    }
#endif
    fwrite(itout64, sizeof(ITEM64), chno, fw);
#if 0
    for(i=0; i < 100; i++)
      dbg("%d] %c%c%c\n", i, itout64[i].ch[0], itout64[i].ch[1], itout64[i].ch[2]);
#endif
  }
  else {
#if NEED_SWAP
    for(i=0; i < chno; i++) {
      swap_byte_4(&itout[i].key);
    }
#endif
    fwrite(itout, sizeof(ITEM), chno, fw);
  }

  if (phr_cou) {
    phridx[phr_cou++]=prbf_cou;
    printf("phrase count:%d\n", phr_cou);

    int ophr_cou = phr_cou;
#if NEED_SWAP
    for(i=0; i < phr_cou; i++)
      swap_byte_4(&phridx[i]);
    swap_byte_4(&phr_cou);
#endif
    fwrite(&phr_cou, sizeof(int), 1, fw);
    fwrite(phridx, sizeof(int), ophr_cou, fw);
    fwrite(phrbuf,1,prbf_cou,fw);
  }

  fclose(fw);

#if 0
  char bzip2[128];
  strcat(strcpy(bzip2, "bzip2 -f -k "), fname_tab);
  system(bzip2);
#endif

  return 0;
}
